Egress gateway modularization

dylandreimerink opened the pull request:
This PR converts the egress gateway into a Hive cell. To facilitate this change the `ipcache.IPCache`, `daemon.CachingIdentityAllocator`, `policy.Repository`, and `certificatemanager` have been modularized as well.

dylandreimerink committed:
certificatemanager,daemon: Modularized the certificate manager

This commit added a Cell for the certificate manager. The cell exposes
two new interfaces `CertificateManager` and `SecretManager` instead
of the manager directly.

The configuration/flags for the manager has been moved into the Cell.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
ipcache,daemon: Modularized IPCache and caching identity allocator

This commit turns three circularly dependent components into Cells,
those being `ipcache.IPCache`, `daemon.CachingIdentityAllocator`, and
`policy.Repository`.

The main objective was to modularize the caching identity allocator but
its dependent on IPCache, which is dependent on the policy repo which
in turn is dependent on the identity allocator, thus making a cycle.

Promises are used to wire the CachingIdentityAllocator and policy repo
since import cycles aren't allowed in Hive.

Daemon was refactored to deal with the increase in passed parameters and
use daemon context for promise awaits (cancelled if start times out)

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
Co-authored-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink committed:
daemon,egressgateway: Modularized the egressgateway manager

This commit turns the egress gateway manager into a Cell. This builds
on the modularization of the CachingIdentityAllocator.

Until now the egress gateway would busy-poll the daemon to see if the
k8s caches are synced. This is now done via a channel on a new object
called `k8s.CacheSyncedChecker` which is shared via Hive to both the
daemon which will modify it and the egress gateway which will subscribe
to it.

This commit also abstracts the egress gateway manager behind an
interface to make it loosely coupled.

Threshold for too many arguments to newDaemon has been passed, so
just use daemonParams directly.

Also respect the start context by cancelling the daemon context if
it times out before start has finished.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
Co-authored-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/ipcache/ipcache.go:
```diff
@@ -110,9 +114,54 @@ type IPCache struct {
+	})
+
+	go func() {
+		policyRepo, err := policyRepoPromise.Await(ctx)
```
Do the awaits from Start() and use HookContext so Await() cancels on start timeout.

joamaki commented on daemon/cmd/policy.go:
```diff
@@ -36,12 +37,51 @@ import (
+		return repo
+	})
+
+	updaterResolver, updaterPromise := promise.New[*policy.Updater]()
```
You can use `promise.Map` here as well. Prefer `Map()` if you can as then the caller doing the Await() will provide the context and we don't need to conjure it here.

joamaki commented on pkg/egressgateway/manager.go:
```diff
@@ -24,25 +25,38 @@ import (
+)
+
+type Manager interface {
+	OnAddEgressPolicy(config PolicyConfig)
```
Are there interface definitions for these we can just embed here? e.g. `subscriber.Node`? Would make the purpose clear.

joamaki commented on pkg/egressgateway/manager.go:
```diff
@@ -77,14 +99,19 @@ func NewEgressGatewayManager(k8sCacheSyncedChecker k8sCacheSyncedChecker, identi
 // then looks up identity by ID from the cached identity allocator and return its labels.
-func (manager *Manager) getIdentityLabels(securityIdentity uint32) (labels.Labels, error) {
+func (mgr *manager) getIdentityLabels(securityIdentity uint32) (labels.Labels, error) {
+	identityAllocator, err := mgr.identityAllocatorPromise.Await(context.TODO())
```
Do the `Await` from `Start()` with HookContext.

joamaki commented on pkg/ipcache/cell.go:
```diff
@@ -0,0 +1,5 @@
+
+import "github.com/cilium/cilium/pkg/hive/cell"
+
+var Cell = cell.Provide(NewIPCache)
```
Consider extracting the interface here. `go doc ipcache.IPCache` gives a good starting point. Also see if you can split the interface up by functionality (e.g. one for lookups, one for modifying etc.).

dylandreimerink replied:
I gave this a shot, we can significantly simplify the public API of the ipcache but it results in quite a significant diff, so I would prefer to do that in a separate PR.

joamaki commented on pkg/wireguard/agent/agent_test.go:
```diff
@@ -80,12 +84,15 @@ func containsIP(allowedIPs []net.IPNet, ipnet *net.IPNet) bool {
+	idResolver, idPromise := promise.New[cache.IdentityAllocator]()
+	idResolver.Resolve(nil)
+	repoResolver, repoPromise := promise.New[*policy.Repository]()
+	repoResolver.Resolve(nil)
```
Hmm, maybe we can add a constructor for resolved promise, or do vararg trick so it can be given optionally (`New[T](v ...T)`).

dylandreimerink replied:
I feel that this is quite niche, it isn't often we have to make pre-resolved promises

joamaki committed:
promise: Add error return value to promise.Map

To also allow rejecting a promise in promise.Map add the error return
value to it.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
